tests/utils/devServer.js

```diff
@@ -0,0 +1,92 @@
+const cliPath = require('./cliPath')
+const path = require('path')
+const getPort = require('get-port')
+const seedrandom = require('seedrandom')
+const execa = require('execa')
+const pidtree = require('pidtree')
+
+// each process gets a starting port based on the pid
+const rng = seedrandom(`${process.pid}`)
+function getRandomPortStart(rng) {
+  const startPort = Math.floor(rng() * 10000) + 10000 // 10000 to avoid collisions with frameworks ports
+  return startPort
+}
+
+let currentPort = getRandomPortStart(rng)
+
+const startServer = async ({ cwd, env = {} }) => {
+  const tryPort = currentPort++
+  const port = await getPort({ port: tryPort })
```

Comment:
Can we use port `0` instead?
It [uses a randomly available port](https://nodejs.org/api/net.html#net_server_listen_port_host_backlog_callback).
Then the port can be retrieved with `server.address()`. Not sure if that would work here though.

Reply:
Tried using random port and it lead to collisions.

Things I can think of that contributed to the collisions:

1. There is a delay between the time `getPort` returns the free port to the time the dev server actually acquires it (since we're spawning it via a child process). That leads to race conditions.
2. The dev server also the underlying framework server on a port which is not randomised yet. For example the create react app server runs on port 3000. Tried using `targetPort` for the `cra` test, but that didn't seem to work.

The current solutions seeds a random port generator for each process based on the process id and increment the port per each test in the same process. Also we add 10000 to the starting port to avoid collisions with frameworks.